File: web/src/api/v2/watchTogetherRoomRead.test.ts
import { afterEach, beforeEach, expect, it, vi } from "vitest";
import { act, cleanup, renderHook, waitFor } from "@testing-library/react";
import {
  setAccessToken,
  setProfileId,
  setProfileToken,
  StaleApiRequestContextError,
} from "@/api/client";
import { getWatchTogetherRoom } from "@/lib/watchTogether";
import { useWatchTogetherRoomConnection } from "@/player/hooks/useWatchTogetherRoomConnection";
const snapshot = {
  room_id: "room",
  phase: "playing",
  playback_state: "paused",
  selection_mode: "host_pick",
  selection_revision: 2,
  selected_file_id: "7",
  selected_library_id: "8",
  code: "ROOM",
  guest_control_policy: "host_only",
  is_paused: true,
  anchor_position_seconds: 3,
  anchor_updated_at: "2026-01-01T00:00:00Z",
  generation: 4,
  member_count: 1,
  host_connected: true,
  self_role: "host",
  self_can_control_transport: true,
  self_can_manage_room: true,
  self_ignore_wait: false,
  members: [
    {
      user_id: "1",
      profile_id: "host",
      display_name: "Host",
      is_host: true,
      is_self: true,
      connected: true,
    },
  ],
};
const response = () =>
  new Response(JSON.stringify({ room: snapshot, room_access_token: "renewed-proof" }), {
    headers: { "Content-Type": "application/json" },
  });
beforeEach(() => {
  localStorage.clear();
  sessionStorage.clear();
  setAccessToken("login");
  setProfileId("host");
  setProfileToken(null);
});
afterEach(() => {
  cleanup();
  vi.unstubAllGlobals();
});
it("reads with header proof and converts exact snapshot identities", async () => {
  const fetch = vi.fn().mockResolvedValue(response());
  vi.stubGlobal("fetch", fetch);
  const result = await getWatchTogetherRoom("room", "original-proof");
  expect(result.room.selected_file_id).toBe(7);
  expect(result.room.members?.[0]?.user_id).toBe(1);
  expect(result.room_access_token).toBe("renewed-proof");
  expect(fetch).toHaveBeenCalledTimes(1);
  expect(fetch.mock.calls[0]![0]).toBe("/api/v2/watch-together/rooms/room");
  expect(new Headers(fetch.mock.calls[0]![1].headers).get("X-Room-Token")).toBe("original-proof");
});
it("refuses unsafe identities and replacement-authority success or failure", async () => {
  const fetch = vi.fn().mockResolvedValue(
    new Response(
      JSON.stringify({
        room: { ...snapshot, selected_file_id: "9007199254740992" },
        room_access_token: "proof",
      }),
      { headers: { "Content-Type": "application/json" } },
    ),
  );
  vi.stubGlobal("fetch", fetch);
  await expect(getWatchTogetherRoom("room", "proof")).rejects.toThrow();
  fetch.mockImplementation(async () => {
    setProfileToken("new");
    return response();
  });
  await expect(getWatchTogetherRoom("room", "proof")).rejects.toThrow();
  fetch.mockImplementation(async () => {
    setProfileToken("next");
    return new Response(null, { status: 409 });
  });
  await expect(getWatchTogetherRoom("room", "proof")).rejects.toBeInstanceOf(
    StaleApiRequestContextError,
  );
});
it("mounted room read treats closed as terminal", async () => {
  class Socket extends EventTarget {
    static OPEN = 1;
    readyState = 0;
    close() {}
    send() {}
  }
  vi.stubGlobal("WebSocket", Socket);
  const fetch = vi.fn().mockImplementation(async (url: string) =>
    String(url).endsWith("/suggestions?limit=100")
      ? new Response(JSON.stringify({ items: [], page: { has_more: false } }), {
          headers: { "Content-Type": "application/json" },
        })
      : new Response(
          JSON.stringify({
            type: "https://siloserver.org/docs/api/v2/problems/conflict",
            title: "Conflict",
            status: 409,
            detail: "Room closed.",
          }),
          { status: 409, headers: { "Content-Type": "application/problem+json" } },
        ),
  );
  vi.stubGlobal("fetch", fetch);
  const { result } = renderHook(() =>
    useWatchTogetherRoomConnection({ roomId: "room", roomToken: "proof" }),
  );
  await waitFor(() => expect(result.current.closedReason).toBe("ended"));
  expect(result.current.room).toBeNull();
});
it("mounted old room read cannot publish after room replacement", async () => {
  class Socket extends EventTarget {
    static OPEN = 1;
    readyState = 0;
    close() {}
    send() {}
  }
  vi.stubGlobal("WebSocket", Socket);
  let resolve!: (r: Response) => void;
  vi.stubGlobal(
    "fetch",
    vi.fn().mockImplementation(async (url: string) =>
      String(url) === "/api/v2/watch-together/rooms/room"
        ? new Promise<Response>((r) => {
            resolve = r;
          })
        : new Response(JSON.stringify({ items: [], page: { has_more: false } }), {
            headers: { "Content-Type": "application/json" },
          }),
    ),
  );
  const { result, rerender } = renderHook(
    ({ id }: { id: string | null }) =>
      useWatchTogetherRoomConnection({ roomId: id, roomToken: "proof" }),
    { initialProps: { id: "room" as string | null } },
  );
  await waitFor(() => expect(resolve).toBeDefined());
  rerender({ id: null });
  await act(async () => {
    resolve(response());
  });
  expect(result.current.room).toBeNull();
});

File: internal/apiv2/watch_together_room_read_test.go
package apiv2

import (
	"context"
	"net/http"
	"strings"
	"testing"

	"github.com/Silo-Server/silo-server/internal/api/handlers"
	"github.com/Silo-Server/silo-server/internal/watchtogether"
)

type fakeRoomRead struct {
	calls int
	err   error
}

func (f *fakeRoomRead) ReadWatchTogetherRoom(_ context.Context, room string, user int, profile, proof string) (watchtogether.Snapshot, string, error) {
	f.calls++
	if user != 1 || profile != "p-owner" || proof != "room-proof" {
		return watchtogether.Snapshot{}, "", &handlers.APIError{Status: 403, Message: "Room proof required."}
	}
	return watchtogether.Snapshot{RoomID: room, Phase: "playing", PlaybackState: "paused", SelectionMode: "host_pick", SelectionRevision: 2, SelectedFileID: new(7), SelectedLibraryID: new(8), GuestControlPolicy: "host_only", AnchorUpdatedAt: "2026-01-01T00:00:00Z", SelfRole: "host", Members: []watchtogether.MemberSummary{{UserID: 1, ProfileID: profile}}}, "renewed-proof", f.err
}
func TestWatchTogetherRoomRead(t *testing.T) {
	f := new(fakeRoomRead)
	deps := pilotDeps(nil, nil)
	deps.WatchTogetherRoomRead = f
	h := NewHandler(deps)
	path := Prefix + "/watch-together/rooms/room-one"
	requireProblem(t, do(t, h, http.MethodGet, path, "", nil), TypeAuthenticationRequired)
	if f.calls != 0 {
		t.Fatal("unauthenticated dispatch")
	}
	requireProblem(t, do(t, h, http.MethodGet, path, "", profileOwner()), TypePermissionDenied)
	headers := profileOwner()
	headers["X-Room-Token"] = "room-proof"
	rec := do(t, h, http.MethodGet, path, "", headers)
	for _, value := range []string{`"selected_file_id":"7"`, `"selected_library_id":"8"`, `"user_id":"1"`, `"room_access_token":"renewed-proof"`} {
		if !strings.Contains(rec.Body.String(), value) {
			t.Fatalf("missing %s: %s", value, rec.Body.String())
		}
	}
	if rec.Code != 200 || rec.Header().Get("Cache-Control") != "no-store" {
		t.Fatalf("%d %v", rec.Code, rec.Header())
	}
	f.err = watchtogether.ErrRoomClosed
	requireProblem(t, do(t, h, http.MethodGet, path, "", headers), TypeConflict)
	f.err = watchtogether.ErrRoomNotFound
	requireProblem(t, do(t, h, http.MethodGet, path, "", headers), TypeNotFound)
	deps.WatchTogetherRoomRead = nil
	requireProblem(t, do(t, NewHandler(deps), http.MethodGet, path, "", headers), TypeDependencyUnavailable)
}
